x-pack/plugins/siem/server/lib/detection_engine/signals/filter_events_with_list.ts

```diff
@@ -0,0 +1,96 @@
+
+import { ListItemArraySchema } from '../../../../../lists/common/schemas/response';
+import { type } from '../../../../../lists/common/schemas/common';
+import { ListClient } from '../../../../../lists/server';
+import { SignalSearchResponse, SearchTypes } from './types';
+import { RuleAlertParams } from '../types';
+
+interface FilterEventsAgainstList {
+  listClient: ListClient;
+  exceptionsList: RuleAlertParams['exceptions_list'];
+  logger: Logger;
+  eventSearchResult: SignalSearchResponse;
+}
+
+export const filterEventsAgainstList = async ({
+  listClient,
+  exceptionsList,
+  logger,
+  eventSearchResult,
+}: FilterEventsAgainstList): Promise<SignalSearchResponse> => {
+  try {
+    if (exceptionsList == null || exceptionsList.length === 0) {
+      return eventSearchResult;
+    }
+
+    // narrow unioned type to be single
+    const isStringableType = (val: SearchTypes) =>
+      ['string', 'number', 'boolean'].includes(typeof val);
+    // grab the signals with values found in the given exception lists.
+    const filteredHitsPromises = exceptionsList.map(async (exceptionItem) => {
+      if (exceptionItem.values == null || exceptionItem.values.length === 0) {
+        throw new Error('Malformed exception list provided');
+      }
+      if (!type.is(exceptionItem.values[0].name)) {
+        throw new Error(`Unsupported list type used, please use one of ${type.keys}`);
+      }
+      if (!exceptionItem.values[0].id) {
+        throw new Error(`Missing list id for exception on field ${exceptionItem.field}`);
+      }
+      // acquire the list values we are checking for.
+      const valuesOfGivenType = eventSearchResult.hits.hits.reduce((acc, searchResultItem) => {
+        const valueField = get(exceptionItem.field, searchResultItem._source);
+        if (valueField != null && isStringableType(valueField) && !acc.has(valueField.toString())) {
+          acc.add(valueField.toString());
+        }
+        return acc;
+      }, new Set<string>());
+
+      const listSignals: ListItemArraySchema = await listClient.getListItemByValues({
+        listId: exceptionItem.values[0].id,
+        type: exceptionItem.values[0].name,
+        value: [...valuesOfGivenType],
+      });
+
+      // create a set of list values that were a hit
+      const badSet = new Set<SearchTypes>(listSignals.map((item) => item.value));
+
+      // do a single search after with these values.
+      // painless script to do nested query in elasticsearch
+      // filter out the search results that match with the values found in the list.
+      const filteredEvents = eventSearchResult.hits.hits.filter((item) => {
+        const eventItem = get(exceptionItem.field, item._source);
+        if (eventItem != null) {
+          return !badSet.has(eventItem);
+        }
+        return false;
+      });
+      const diff = eventSearchResult.hits.hits.length - filteredEvents.length;
+      logger.debug(`Lists filtered out ${diff} events`);
```

Comment:
I like this. Wonder if this is a metric we can add to be accumulated for all exceptions (not just lists) to report to the user how many events are filtered.